Fix CPU bitwise shifts for out-of-limit shift values

bjhargrave opened the pull request:
Negative shift values and positive shift values greater than the bit size of the dtype (limit `0..bits`) now yield expected results which are consistent with numpy.

Left shift with an out-of-limit shift value result in a value of `0`. Right shift with an out-of-limit shift value results in a value of `-1` for negative inputs and `0` for non-negative inputs (sign preserving).

Fixes https://github.com/pytorch/pytorch/issues/70904



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

bjhargrave committed:
Fix CPU bitwise shifts for out-of-limit shift values

Negative shift values and positive shift values greater than the
bit size of the dtype (limit 0..bits) now yield expected results
which are consistent with numpy.

Left shift with an out-of-limit shift value result in a value of 0.
Right shift with an out-of-limit shift value results in a value of -1
for negative inputs and 0 for non-negative inputs (sign preserving).

Fixes https://github.com/pytorch/pytorch/issues/70904

Signed-off-by: BJ Hargrave <hargrave@us.ibm.com>

jgong5 commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -314,12 +314,21 @@ void bitwise_xor_kernel(TensorIteratorBase& iter) {
     cpu_kernel_vec(iter,
         [](scalar_t a, scalar_t b) -> scalar_t {
-          return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+          if ((b >= 0) && (b < shift)) {
+            return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+          }
+          return 0;
         },
-        [](Vectorized<scalar_t> a, Vectorized<scalar_t> b) {
-            return a << b;
+        [&zero_vec, &shift_vec](vec_t a, vec_t b) -> vec_t {
+          auto mask = (b >= zero_vec) & (b < shift_vec);
+          a = vec_t::blendv(zero_vec, a, mask);
+          return a << b;
```
May I know what is the problem for vectorized implementation here? The `_mm256_srav_epi32` and `_mm256_sllv_epi32` etc. should already do the right things and we don't have to do extra masking here, right?

albanD replied:
What do you mean by "the right thing"? These are UB from the standard iirc.

jgong5 replied:
I mean the semantics of the intrinsics/instructions already guarantee the expected behavior, e.g., when shifting left with outside-range count, zero is returned and when shifting right, sign flag is preserved, no need to do line 329-330 (and similarly for right shift). Or did I miss anything here?

jgong5 replied:
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_srav_epi32
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_sllv_epi32

albanD replied:
Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

bjhargrave replied:
The nightly macos build is compiled w/o AVX2 and it fails the new test. So a fix is necessary for other CPUs.

An alternate solution, which assumes the SIMD implementations of `operator<<` and `operator>>` are correct, would be to fix the `operator<<` and `operator>>` implementations in `vec_base.h` for when SIMD implementation is not supported. I will try this.

bjhargrave replied:
Turns out there was a bug in `vec256` for the [`operator>>` implementation for `int64`](https://github.com/pytorch/pytorch/blob/ba4fb9b6ad0b3a06ae1b15a170cea6f4223bda79/aten/src/ATen/cpu/vec/vec256/vec256_int.h#L1482-L1497). Since there is no shift right arithmetic vector instruction, the implementation emulated the instruction. However the implementation did not work if the shift value was out-of-limit (<0 or >64). When the shift value was out-of-limit, subtracting the shift value from 64 to compute the sign shift will result in a bad sign shift value and bad sign extension value for negative input.

I have added a new commit to the PR which fixes this.

jgong5 replied:
> Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

The behavior should be consistent on x86 CPUs in general. In general, I think it is better to address the problem in the vectorized implementation to avoid unnecessary guards where possible.

albanD reviewed (approved):
SGTM

bjhargrave committed:
Fix operator>> for int64 vector in vec256

There is no vector instruction for shift right arithmetic for int64.
The operator>> implementation emulates this through other vector
instructions. It has been fixed to properly handle out-of-limit
shift values so that shift values <0 and >64 are set to 64 which
results in a value of -1 for negative inputs and 0 for non-negative
inputs (sign preserving).

Fixes https://github.com/pytorch/pytorch/issues/70904

Signed-off-by: BJ Hargrave <hargrave@us.ibm.com>

albanD reviewed (commented):
Thanks everyone!

@pytorchbot merge

albanD commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -f "ROCM failures are irrelevant"
